pypeit/images/rawimage.py

```diff
@@ -1125,10 +1132,25 @@ def subtract_scattlight(self):
             msgs.warn("The scattered light has already been subtracted from the image!")
             return
 
+        if scattlight.scattlight_param is None:
+            msgs.warn("Scattered light parameters are not set. Cannot perform scattered light subtraction.")
+            return
+
         # Loop over the images
         for ii in range(self.nimg):
-            binning = self.detector[0].binning
-            scatt_img = self.spectrograph.scattered_light(self.image[ii, ...], binning)
+            # Calculate a model specific for this frame
+            spatbin = parse.parse_binning(self.detector[0]['binning'])[1]
+            # TODO :: Need to pass in the fitstbl here... currently this is a dummy value
+            tbl = dict(binning=self.detector[0]['binning'], dispname='BL')
+            pad = scattlight.pad // spatbin
+            offslitmask = slits.slit_img(pad=pad, initial=True, flexure=None) == -1
+            scatt_img, _, success = self.spectrograph.scattered_light(self.image[ii, ...], offslitmask, tbl)
+            # If failure, revert back to the Scattered Light calibration frame model parameters
+            if not success:
+                msgs.warn("Scattered light model failed - using predefined model parameters")
+                # Use predefined model parameters
+                scatt_img = self.spectrograph.scattered_light_model(scattlight.scattlight_param, self.image[ii, ...])
```

Comment:
My understanding of the current flow is:
 - If scattered-light frames are provided, fit a scattered light model to them.  If they are not, scattered light is never subtracted.
 - For the science frames, the code will try to independently model the scattered light in them (only if scattered light frames are provided).  If that fails, use the scattered light model from the scattered-light frames.

Is this right?  If not, can you describe the flow?

If so, can we enable the scattered-light model to be subtracted from the science data regardless of whether or not you provide scattered-light frames?

Also, if I understand the source of the scattered light for KCWI, at least, (i.e., its intensity is proportional to the light flux through the system), then is it appropriate to subtract the model from the science frames given the significant difference in flux?

Reply:
Your understanding of the control flow is correct. But, I'm still deciding what is the best approach. I think it's best to general the model parameters and always use those, unless the user explicitly wants to calculate the scattered light model from each individual frame. I will change the flow such that it's like that, and document it here so the flow is clearer.

> If so, can we enable the scattered-light model to be subtracted from the science data regardless of whether or not you provide scattered-light frames?

Yes, that is possible, and I think that may even be better in some cases... still testing...

> Also, if I understand the source of the scattered light for KCWI, at least, (i.e., its intensity is proportional to the light flux through the system), then is it appropriate to subtract the model from the science frames given the significant difference in flux?

That is correct. This is precisely what is currently being done. For each setup there's a set of model parameters that:
(1) take the light distribution of the current frame (e.g. a science frame)
(2) based on this light distribution on the detector (and the pre-defined model parameters), calculate the scattered light image
(3) subtract this scattered light image from the current frame.